### ISSUE TITLE
[Bug] FA3 varlen + cuda-graph shared pool: reversed bs capture order causes deterministic IMA (root cause for #22096, #14624)

### ISSUE THREAD

**Issue by hnyls2002** [sgl-project/sglang#26532]
> ## Resolved (closed 2026-06-08) — original hypothesis was incorrect
>
> The capture-order / FA3-workspace-slot theory in this issue is **wrong** (see the closing comment for the detailed why). The real root cause for this `spec + topk>1 + page_size>1 + fa3` config was an **out-of-bounds scatter / under-allocation in the EAGLE draft-decode CUDA-graph metadata**, fixed by:
>
> - #27360 — fa3 EAGLE draft-decode expand `page_table` scatter OOB (topk>1 + page>1)
> - #27338 / #27460 — EAGLE draft CUDA-graph `kv_indices` under-allocation (topk>1)
> - #27486 — defensive guards for EAGLE draft KV indexing
>
> The reversed cuda-graph capture order is **intentional and remains unchanged on `main`**. #22096 / #14624 share the same trigger config and are covered by the same fixes.

---

## Appendix — original hypothesis (DISPROVEN, kept for history)

The original claims, struck through:

- ~~FA3 `flash_attn_with_kvcache` varlen workspace slot ends up sized by the LAST capture into the shared cuda-graph mempool.~~
- ~~Reversed bs capture order (smallest bs captured last) under-sizes that slot, so a larger-bs replay over-reads -> IMA / NaN / `vectorized_gather` assert (three faces of the same over-read).~~
- ~~Fix = switch `CudaGraphRunner` capture to ascending bs order.~~

**Why wrong:** the minimal repro frees the graph-input `q` / `cuq` each loop iteration — a graph-input-liveness UB (flagged by @DarkSharpness), not a workspace-sizing bug — and `--fix` mode adds a keep-alive hack *only* in fix mode, confounding capture order with input liveness. Reversed capture (largest bs first) is the correct pool-sharing pattern. The real bug was the EAGLE draft-decode metadata OOB listed above.

<details>
<summary>Original full issue text + repro script (kept for history)</summary>

## Summary

FA3 `flash_attn_with_kvcache` on the VARLEN multi-query path (`q_len > 1`, `VarlenDynamicPersistentTileScheduler`) allocates a per-call workspace whose slot in a shared cuda-graph mempool ends up sized by the LAST capture into that slot. SGLang's `CudaGraphRunner` captures bs's in **reversed** order (`python/sglang/srt/model_executor/cuda_graph_runner.py:824-829`, comment: *"Reverse the order to enable better memory sharing across cuda graphs"*), so the smallest bs is captured last and the slot is sized for the smallest graph. Replaying a larger-bs graph then over-reads / over-writes past the slot, surfacing as one of three layout-dependent faces:

- **IMA** -- over-read lands on an unmapped page (hardware fault).
- **NaN / wrong logits** -- over-read lands on mapped scheduler state, silently corrupts.
- **`vectorized_gather` assert** -- corrupted indices fail a downstream bounds check.

All three are variants of the *same* over-read, which explains the three-headed flakiness in production CI:

- #22096 (open) -- `test_eagle_infer_b` CI flakiness (immediate motivation).
- #14624 (closed, *"Hard to reproduce"*) -- same trigger config (spec + topk>1 + page>1 + fa3); this issue provides the missing deterministic repro and should justify reopening.
- #21112 (open) -- FA3 + piecewise cuda-graph IMA at `warmup_compile`; possibly the same kernel-level root cause on a different capture surface.

## Deterministic minimal repro

92-line self-contained script. Needs an FA3 build (`sgl_kernel.flash_attn`) on SM90 (H100/H200).

```python
#!/usr/bin/env python3
"""
Deterministic repro / fix-verifier for the FA3 (hopper) varlen cuda-graph IMA.

FA3 flash_attn_with_kvcache on the VARLEN multi-query path (q_len>1,
VarlenDynamicPersistentTileScheduler) allocates a per-call workspace that is
reused at the same pool offset across the per-bs cuda graphs. If the SMALLEST bs
is captured LAST (decreasing/`reversed` order, as SGLang's CudaGraphRunner does),
that offset's slot is sized for the smallest graph, so replaying a larger-bs
graph overflows it -> illegal memory access (flaky in prod: crash vs NaN logits
depending on what the over-read hits).

  (default) : capture bs in DECREASING order (5..1) == current SGLang behavior -> IMA CRASH
  --fix     : capture bs in INCREASING order (1..5) == proposed fix            -> NO-CRASH + CORRECT

Run:  CUDA_VISIBLE_DEVICES=0 python3 fa3_cudagraph_repro.py [--fix]
Needs: an FA3 build (sgl_kernel.flash_attn) on an SM90 GPU (H100/H200).
"""
import argparse
import torch
from sgl_kernel.flash_attn import flash_attn_with_kvcache

ap = argparse.ArgumentParser()
ap.add_argument("--fix", action="store_true",
                help="capture bs in INCREASING order (proposed fix); without this flag reproduces the IMA crash")
args = ap.parse_args()

DEV, DT = "cuda", torch.float16
NH, HD, PAGE, QLEN, NPAGES = 32, 128, 256, 64, 80   # QLEN>1 => varlen persistent path
torch.manual_seed(0)
kc = torch.randn(NPAGES, PAGE, NH, HD, dtype=DT, device=DEV)
vc = torch.randn(NPAGES, PAGE, NH, HD, dtype=DT, device=DEV)

def attn(q, page_table, cache_seqlens, cu_seqlens_q):
    return flash_attn_with_kvcache(
        q=q, k_cache=kc, v_cache=vc, page_table=page_table,
        cache_seqlens=cache_seqlens, cu_seqlens_q=cu_seqlens_q, max_seqlen_q=QLEN,
        softmax_scale=HD**-0.5, causal=False, return_softmax_lse=False, num_splits=0,
    )

# Match the original IMA-reproducing layout: every bs allocates fresh
# (q, pt, cs, cuq) inside the capture loop and only (g, o, pt, cs) is retained;
# q and cuq are dropped at iteration end so the caching allocator does not
# pre-map pages adjacent to the FA3 per-call workspace.
# In --fix mode only, we additionally hold bs=4's q and cuq so the caching
# allocator cannot reuse their GPU addresses for later bs's q/cuq. Otherwise
# the captured bs=4 graph would read q content overwritten by bs=5's q and
# the eager comparison would show a spurious non-zero diff. In default mode
# the IMA crashes before reaching the eager comparison anyway.
pool = torch.cuda.graph_pool_handle()
graphs = {}
q4 = None
cuq4 = None
bs_list = [1, 2, 3, 4, 5] if args.fix else [5, 4, 3, 2, 1]
for bs in bs_list:
    q   = torch.randn(bs * QLEN, NH, HD, dtype=DT, device=DEV)
    pt  = torch.zeros(bs, 1, dtype=torch.int32, device=DEV)
    cs  = torch.zeros(bs, dtype=torch.int32, device=DEV)
    cuq = torch.arange(0, bs * QLEN + 1, QLEN, dtype=torch.int32, device=DEV)
    s = torch.cuda.Stream(); s.wait_stream(torch.cuda.current_stream())
    with torch.cuda.stream(s):
        for _ in range(3): o = attn(q, pt, cs, cuq)
    torch.cuda.current_stream().wait_stream(s); torch.cuda.synchronize()
    g = torch.cuda.CUDAGraph()
    with torch.cuda.graph(g, pool=pool):
        o = attn(q, pt, cs, cuq)
    if bs == 4 and args.fix:
        q4, cuq4 = q, cuq
    graphs[bs] = dict(g=g, o=o, pt=pt, cs=cs)
print(f"[capture] order = {bs_list}  ({'increasing (fix)' if args.fix else 'reversed (bug)'})")

# Replay bs=4 with realistic non-zero prefix lengths.
bs = 4
graphs[bs]["cs"].copy_(torch.tensor([220, 230, 240, 253], dtype=torch.int32, device=DEV))
graphs[bs]["pt"][:, :1].copy_(torch.tensor([[1], [2], [3], [4]], dtype=torch.int32, device=DEV))
try:
    graphs[bs]["g"].replay(); torch.cuda.synchronize()
    out = graphs[bs]["o"].float().clone()
    nan = bool(torch.isnan(out).any())
    if args.fix:
        # Eager ref using the same (q, pt, cs, cuq) the captured graph references.
        ref  = attn(q4, graphs[bs]["pt"], graphs[bs]["cs"], cuq4).float()
        diff = (out - ref).abs().max().item()
        print(f"[RESULT] NO-CRASH  nan={nan}  max_abs_diff_vs_eager={diff:.4g}  "
              f"-> {'CORRECT' if (not nan and diff == 0) else 'WRONG/CORRUPT'}")
    else:
        # Default (reversed) mode is expected to CRASH; if we land here the
        # IMA did not fire this run. q-buffer reuse corrupts the captured
        # input so eager diff would be a spurious artifact, not a bug signal.
        print(f"[RESULT] NO-CRASH  nan={nan}  (unexpected: IMA did not trigger this run)")
except Exception as e:
    print(f"[RESULT] CRASH: {str(e).splitlines()[0]}")
```

Run:

```bash
python3 fa3_cudagraph_repro.py          # reversed [5,4,3,2,1] (current behavior) -> CRASH (IMA)
python3 fa3_cudagraph_repro.py --fix    # increasing [1,2,3,4,5] (proposed fix)   -> NO-CRASH + diff=0
```

Verified deterministic on H200 with sgl_kernel 0.4.2.post2 and 0.4.3, 3/3 runs each.

## What's verified

- `[5,4,3,2,1]` reversed capture -> deterministic IMA on bs=4 replay.
- `[1,2,3,4,5]` increasing capture -> bit-exact correct output vs eager (`diff=0`).
- An earlier layout variant of the script reproduced the silent-corruption face (`max_abs_diff_vs_eager=0.9641`) with the same `[5,4,3,2,1]` capture, confirming IMA and NaN/wrong-output are layout variants of the same over-read.

## Caveat about the repro

The script holds bs=4's `q` and `cuq` in `--fix` mode (`if bs == 4 and args.fix: q4, cuq4 = q, cuq`) to suppress a separate q-buffer reuse artifact that exists in this minimal script but NOT in production. Production `CudaGraphRunner.InputBuffers` uses persistent input buffers shared across all per-bs graphs, so a captured graph's input addresses are never silently overwritten by a later capture. The proposed SGLang-side fix does NOT require this hack.

## Suggested fix and tradeoff

`python/sglang/srt/model_executor/cuda_graph_runner.py:824-829`:

```python
# Reverse the order to enable better memory sharing across cuda graphs.
capture_range = (
    tqdm.tqdm(list(reversed(self.capture_bs)))
    if get_tensor_model_parallel_rank() == 0
    else reversed(self.capture_bs)
)
```

Switching to ascending bs order makes the FA3 per-call workspace slot sized for the LARGEST bs, so any subsequent replay (always <= max_bs) fits. The existing `reversed(...)` is an intentional memory-sharing optimization (per the inline comment), so the fix should:

1. Switch to ascending order on the FA3 backend (or unconditionally), AND
2. Measure the memory-sharing impact for production configs and document it.

## Validation path

This minimal repro proves the kernel-level bug class is real and capture-order-triggered. Whether the SGLang-side fix fully resolves CI flakiness should be validated by running `test_eagle_infer_b TestEAGLEServerAdditional` N x N (before/after) on production CI.

## Environment

- SGLang main @ `24bcb37efb`
- sgl_kernel 0.4.2.post2 and 0.4.3 (both reproduce)
- H200 (SM90), CUDA 12.x
- Trigger config (matches `test_eagle_infer_b TestEAGLEServerAdditional`):
  - attention-backend=fa3, page-size=256, topk=5, num-steps=8, dtype=float16, cuda-graph-max-bs=5
  - Llama-2-7b-chat target + lmsys/sglang-EAGLE-llama2-chat-7B draft (in production)
  - minimal repro uses synthetic kc/vc + varlen multi-query directly against `sgl_kernel.flash_attn`

</details>

**Comment by DarkSharpness** [comment]
There's undefined behavior in this example. The q allocated in the first iteration of the loop is also the input of the first CUDA graph. However, in the second iteration, the first q will be freed because we allocate a fresh new q. For torch CUDA graph, we must ensure that all the tensors allocated outside and used inside should be kept alive.

**Comment by hnyls2002** [comment]
Closing — the capture-order hypothesis here was a red herring.

Two problems with the original repro: (1) default (bug) mode frees the graph-input `q` / `cuq` every loop iteration, so the IMA is the classic "freed graph-input tensor reused at a stale address" UB (as @DarkSharpness noted), not workspace sizing; (2) `--fix` mode adds a `q4` / `cuq4` keep-alive *only* in fix mode, so the before/after compares `reversed + no-keepalive` vs `ascending + keepalive` — capture order and input liveness are confounded. Reversed capture (largest bs captured first) is the correct pool-sharing pattern and is unchanged on `main`.

The real root cause for this config (`spec + topk>1 + page_size>1 + fa3`) was an out-of-bounds scatter / under-allocation in the EAGLE draft-decode CUDA-graph metadata, fixed by:

- #27360 — fa3 EAGLE draft-decode expand `page_table` scatter OOB (topk>1 + page>1)
- #27338 / #27460 — EAGLE draft CUDA-graph `kv_indices` under-allocation (topk>1)
- #27486 — defensive guards for EAGLE draft KV indexing

These match the IMA / NaN / `vectorized_gather` assert symptom triad and the page_size=256 flakiness (contiguous draft slots usually share one page, so the OOB only intermittently crosses a page boundary). #22096 and #14624 share the same trigger config and are covered by the same fixes.
